Pip 6 and Django 1.8 and 1.9 compatability

richardxia opened the pull request:
@lucaswiman 

This PR brings this library up to date with Django 1.9 as well as pip 6+. The solution for fixing compatibility with pip 6.0.0 may have broken compatibility with pip < 1.5, which is why I bumped the major version number of this library. AFAICT, the actual behavior of the library should not have changed, but I can't guarantee that my changes work under Django 1.5 and 1.6, which this library previously supported.

The main changes required for Django 1.8 and 1.9 compatibility involve fixing how the `TEMPLATE_STRING_IF_INVALID` setting is patched, since it was change in both 1.8 and 1.9. I wrote wrapper functions for patching and retrieving the value of the `string_if_invalid` that branch based on the Django version, leaving most of the core code the same.

In addition, there were a couple minor bugs that I found and fixed, which I can comment on inline.

Lastly, I added a `Makefile` that helps with testing, and I added `tox` to allow us to test Django 1.7, 1.8, and 1.9.

Commit:
Pass in required session keyword argument.

Commit:
Add in tox and Makefile, and fix up requirements.

Commit:
Fix import path issues in later versions of Django.

Commit:
Patch builtin tags in a version agnostic way.

Commit:
Patch string_if_invalid in a version-agnostic way.

Commit:
Actually attempt to return original render.

Commit:
Get value of invalid template string in version agnostic way.

Commit:
Fix test patches.

Commit:
Fix ifdef tag for non-identifier expressions.

Commit:
Do not guarantee Django 1.5/1.5 support. Bump version number.

Commit:
Update instructions for testing.

richardxia commented on pedant/decorators.py:
```diff
@@ -197,7 +226,7 @@ def log_debug_render(*args, **kwargs):
     def log_render(*args, **kwargs):
         try:
-            variable_node_render(*args, **kwargs)
+            return variable_node_render(*args, **kwargs)
```
This seemed to be an existing bug. Unlike the `debug` version of this function (just above), this one previously wasn't returning the actual result if there was no UnicodeDecodeError. This was probably just an oversight and wasn't caught by the tests because the tests only run the `debug` version on Django <= 1.8. Django 1.9 drops the `debug` versions of the template objects, which is how I noticed this bug.

lucaswiman replied:
Nice catch.

richardxia commented on pedant/templatetags/pedant_tags.py:
```diff
@@ -49,6 +49,8 @@ def ifdef(parser, token):
     # {% ifdef ... %}
     bits = token.split_contents()[1:]
+    if len(bits) > 1:
+        raise TemplateSyntaxError('%r is not an identifier.' % token)
```
Without this change, the `{% ifdef %}` was failing the test that checks that you can't pass arbitrary expressions into `{% ifdef %}`. It was failing on all the versions of Django I tested this on, from 1.5 through 1.9, but I'm not sure why it was passing for you a couple years ago. I'm making the assumption that all inputs with more than one argument cannot be a single identifier, which I can't think of a counterexample for.

richardxia commented on pedant/tests.py:
```diff
@@ -485,7 +498,8 @@ def render1():
         self.assertEqual(logged_render, original_render)
 
-    @override_settings(TEMPLATE_STRING_IF_INVALID='Fixed String')
+    @skipIf(django.VERSION >= (1, 8), 'Not an issue as of Django 1.8')
```
These tests check that Django's cached version of `TEMPLATE_STRING_IF_INVALID` is overwritten when applying the patch decorators. Django no longer caches `TEMPLATE_STRING_IF_INVALID` onto `django.template.base.invalid_var_format_string`, so these tests are no longer relevant. It _does_ cache it onto the template Engine, but version of the patcher that I wrote for 1.8 and 1.9 should patch the underlying Engine directly instead of the setting.

lucaswiman commented on Makefile:
```diff
@@ -0,0 +1,104 @@
+# Makefile for packaging and testing this app. Should follow make contract
```
I think this makefile is mostly superfluous. The tests pass fine just running `tox`. The only benefits of the makefile AFAICT are:
- Combining coverage information between all the different tox environments.
- A target for tagging the releases.
- It functions in environments which don't have tox already installed.

WDYT about just getting rid of it? If we do want additional boilerplate for running the test suite, it would ideally be a `.travis.yml` file.

richardxia replied:
Sounds good. Removed in febc7a7e71f2c036ab94efd61a89affac39d16f2.

lucaswiman commented on pedant/decorators.py:
```diff
@@ -206,8 +235,13 @@ def log_render(*args, **kwargs):
-    patch_debug = patch.object(DebugVariableNode, 'render', log_debug_render)
-    return lambda f: patch_base(patch_debug(f))
+    patch_all = patch_base
+    if django.VERSION[:3] < (1, 9):
```
Nit. Tuples are ordered lexicographically starting from the zeroth index, so I think `if django.VERSION < (1, 9):` is correct and more concise.

richardxia replied:
Ah, good point. For the version tuples I'm using, the comparisons should work fine. I was initially concerned about the 4th tuple item having a value that could be one of `['alpha', 'beta', 'rc', 'final']`, where lexicographic comparisons could be wrong ("release candidate" should be smaller than "final"). However, because I don't use the 4th item, the comparisons should be equivalent. Updated in 5236cbedab08e7196f6b49f19d4ee7b044447697

Commit:
Remove Makefile in favor of just running tox directly

Commit:
Add coverage to requirements-dev.txt. Reduce coverage requirement.

Commit:
Add a bit more test coverage.

Commit:
Do comparison on full version tuple.

richardxia commented:
@lucaswiman, I made one more change that was not related to one of your comments. After I removed the `Makefile`, it allowed `tox` to print out warnings in color, and I noticed that I wasn't actually running code coverage because `coverage` wasn't included in `requirements-dev.txt`. After adding it back in, I realized that I was failing the 100% code coverage requirement, but because most of the uncovered paths are due to Django version branching, it's not possible for a single tox environment to achieve 100% test coverage. The 1.9 environment is particularly bad because it skips the most lines of code.

I made the code coverage requirement environment specific, setting it to 90% for 1.7 and 1.8, and 80% for 1.9: 6fe8bc7

I also did attempt to add a bit more code coverage to the `ifdef` stuff I added: 2e44515.

PTAL again.

lucaswiman commented on tox.ini:
```diff
@@ -6,7 +6,9 @@ envlist = py27-{1.7,1.8,1.9}
-  NOSE_COVER_MIN_PERCENTAGE=100 
+  1.7: NOSE_COVER_MIN_PERCENTAGE=90
+  1.8: NOSE_COVER_MIN_PERCENTAGE=90
+  1.9: NOSE_COVER_MIN_PERCENTAGE=80
```
This is a bit hacky, but and it would probably make sense to combine the coverage reports in the travis.yml file at a later date. I'm happy to add that to this library.

richardxia replied:
Cool, thanks, that sounds like the better solution.

lucaswiman commented:
LGTM.

richardxia commented:
Thanks for reviewing, @lucaswiman!
